src/colmap/util/threading_ctrl.h

```diff
@@ -0,0 +1,259 @@
+// POSSIBILITY OF SUCH DAMAGE.
+
+#pragma once
+
+#include "colmap/util/logging.h"
+#include "colmap/util/timer.h"
+
+#include <atomic>
+#include <climits>
+#include <functional>
+#include <future>
+#include <list>
+#include <queue>
+#include <thread>
+#include <type_traits>
+#include <unordered_map>
+
+namespace colmap {
+
+#ifdef __clang__
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wkeyword-macro"
+#endif
+
+#ifdef __clang__
+#pragma clang diagnostic pop  // -Wkeyword-macro
+#endif
+
+// Reimplementation of threading with thread-related functions outside
+// controller Following util/threading.h
+// BaseController that supports templating in ControllerThread
+class BaseController {
+ public:
+  BaseController();
+  virtual ~BaseController() = default;
+
+  enum {
+    STARTED_CALLBACK = INT_MIN,
+    FINISHED_CALLBACK,
+
+    // thread-related callbacks from here, only useful in the Thread class,
+    // empty as default
+    LOCK_MUTEX_CALLBACK,
+    BLOCK_IF_PAUSED_CALLBACK,
+    SIGNAL_SETUP_CALLBACK,
+    CHECK_VALID_SETUP_CALLBACK,
+  };
+
+  // Set callbacks that can be triggered within the main run function.
+  void AddCallback(int id, const std::function<void()>& func);
+
+  // This is the main run function to be implemented by the child class. If you
+  // are looping over data and want to support the pause operation, call
+  // `BlockIfPaused` at appropriate places in the loop. To support the stop
+  // operation, check the `IsStopped` state and early return from this method.
+  virtual void Run() = 0;
+
+  // wrapped function for threading
+  void RunFunc();
+
+  // check if the thread is stopped
+  void SetCheckIfStoppedFunc(const std::function<bool()>& func);
+  bool IsStopped() const;
+  // BlockIfPaused + IsStopped()
+  bool CheckIfStopped();
+
+  // To be called from inside the main run function. This blocks the main
+  // caller, if the thread is paused, until the thread is resumed.
+  void BlockIfPaused();
+
+  // To be called from outside. This blocks the caller until the thread is
+  // setup, i.e. it signaled that its setup was valid or not. If it never gives
+  // this signal, this call will block the caller infinitely. Check whether
+  // setup is valid. Note that the result is only meaningful if the thread gives
+  // a setup signal.
+  bool CheckValidSetup();
+
+  // test if setup is called
+  bool SetupCalled() const { return setup_; }
+
+ protected:
+  // Register a new callback. Note that only registered callbacks can be
+  // set/reset and called from within the thread. Hence, this method should be
+  // called from the derived thread constructor.
+  void RegisterCallback(int id);
+
+  // Call back to the function with the specified name, if it exists.
+  void Callback(int id) const;
+
+  // Signal that the thread is setup. Only call this function once.
+  void SignalValidSetup();
+  void SignalInvalidSetup();
+
+ private:
+  bool setup_ = false;
+  bool setup_valid_ = false;
```

Comment:
Can't this go into the `ControllerThread` instead? By adding an additional callback `signal_valid_setup_fn` similar to `check_if_stopped_fn`.

Reply:
These two flags are not necessarily related to threading. For the ``SignalValidSeup`` and ``SignalInvalidSetup`` we need a boolean variable in the controller for checking.

Reply:
IIUC `CheckValidSetup` exists only because we can't easily propagate errors in threading. It's used only in nested threads in feature extraction/matching. When there is no threading involved, it is useless - one should rather return a flag or throw an exception. So I think that the booleans should be part of the threading class.

Regarding the signaling callbacks: if we make them explicit callbacks (like `signal_valid_setup_fn`) then we can merge them into a single one that accepts a boolean (setup is valid or not).

Reply:
With the new design this method is not needed in the controller anymore